src/Xrefcheck/Verify.hs

```diff
@@ -81,230 +87,324 @@ toVerifyRes = VerifyResult . either one (\() -> [])
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+            Right (url, option) ->
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+
+      let maxTime = Time @Second $
+            unTime vcExternalRefCheckTimeout * timeoutFrac
+
+      mres <- liftIO (timeout maxTime $ void reqLink) `catch`
+        (either throwError (\() -> return (Just ())) . interpretErrors)
+      maybe (throwError $ ExternalResourceSomeError "Response timeout") pure mres
 
     isAllowedErrorCode = or . sequence
-        -- We have to stay conservative - if some URL can be accessed under
-        -- some circumstances, we should do our best to report it as fine.
-        [ (403 ==)  -- unauthorized access
-        , (405 ==)  -- method mismatch
-        ]
+      -- We have to stay conservative - if some URL can be accessed under
+      -- some circumstances, we should do our best to report it as fine.
+      [ (403 ==)  -- unauthorized access
+      , (405 ==)  -- method mismatch
+      ]
 
     interpretErrors = \case
-        JsonHttpException _ -> error "External link JSON parse exception"
-        VanillaHttpException err -> case err of
-            InvalidUrlException{} -> error "External link URL invalid exception"
-            HttpExceptionRequest _ exc -> case exc of
-                StatusCodeException resp _
-                    | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
-                    | otherwise -> Left $ ExternalResourceUnavailable (responseStatus resp)
-                other -> Left . ExternalResourceSomeError $ show other
+      JsonHttpException _ -> error "External link JSON parse exception"
+      VanillaHttpException err -> case err of
+        InvalidUrlException{} -> error "External link URL invalid exception"
+        HttpExceptionRequest _ exc -> case exc of
+          StatusCodeException resp _
+            | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
+            | otherwise -> Left $
+                ExternalHttpResourceUnavailable (responseStatus resp)
+          other -> Left . ExternalResourceSomeError $ show other
+
+    checkFtp :: URI -> Bool -> ExceptT VerifyError IO ()
+    checkFtp uri secure = do
+      -- get authority which stores host and port
+      authority <- case uriAuthority uri of
+        Right a -> pure a
+        Left _ -> throwError ExternalResourceInvalidUri
+      let host = T.unpack . unRText $ authHost authority
+      let port :: Int = fromIntegral . fromMaybe 21 $ authPort authority
+      let pieces = uriPath uri
+      -- in case there aro no pieces - smth wrong with url
+      unless (isJust pieces) $
+        throwError ExternalResourceInvalidUri
+      let (trailing, pathList) = fromJust pieces
+      -- in case there are trailing slash - smth wrong with url
+      when trailing $
+        throwError ExternalResourceInvalidUri
```

Comment:
Same here: In this function, we're throwing `ExternalResourceInvalidUri` in three places, but we should be throwing a `ExternalResourceInvalidURL` instead.

Also, since we know _why_ the URI is not a valid URL (no authority, no path, etc), the error message should include this piece of information.